README.rst
===========
pyspotlight
===========

is a thin python wrapper around `DBpedia Spotlight`_'s `REST Interface`_.

This package is tested against DBpedia Spotlight version 0.7.
As long as there are no major API overhauls, this wrapper might also
work with future versions. If you encounter a bug with a newer DBpedia Spotlight version,
feel free to create an issue here on github.

Note that we're trying to track DBpedia Spotlight release version numbers, so you can
easily see which pyspotlight version has been tested with which Spotlight
release. For example, all pyspotlight 0.6.x releases are compatible with
Spotlight 0.6.x, etc. While we aim for backwards-compatibility with older
Spotlight releases, it is not guaranteed. If you're using an older Spotlight
version, you may need to use an older pyspotlight version as well.

.. _`DBpedia Spotlight`: http://www.dbpedia-spotlight.org/faq
.. _`REST Interface`: http://www.dbpedia-spotlight.org/api

Installation
============

The newest stable release can be found on the `Python Package Index (PyPI) <https://pypi.python.org/pypi>`__.

Therefore installation is as easy as::

    pip install pyspotlight

Older releases can be installed by specifying a version::

    pip install pyspotlight~=0.6.1

Requirements for installation from source/github
================================================

This module has been tested with Python 2.7 and Python 3.5.

As long as you use the ``setup.py`` for the installation
(``python setup.py install``), you'll be fine because Python takes care of the
dependencies for you.

If you decide not to use the ``setup.py`` you will need the ``requests``
library.

All of these packages can be found on the `Python PackageIndex`_ and easily
installed via either ``easy_install`` or, `the recommended`_, ``pip``.

Using ``pip`` it is especially easy because you can just do this::

    pip install -r requirements.txt

and it will install all package dependencies listed in that file.

.. _`Python PackageIndex`: http://pypi.python.org/
.. _`the recommended`: http://stackoverflow.com/questions/3220404/why-use-pip-over-easy-install

Usage
=====

Usage is simple and easy, just as the API is::

    >>> import spotlight
    >>> annotations = spotlight.annotate('http://localhost/rest/annotate',
    ...                                  'Your test text',
    ...                                  confidence=0.4, support=20)

This should return a list of all resources found within the given text.
Assuming we did this for the following text::

    President Obama on Monday will call for a new minimum tax rate for individuals making more than $1 million a year to ensure that they pay at least the same percentage of their earnings as other taxpayers, according to administration officials.

We might get this back::

    >>> spotlight.annotate('http://localhost/rest/annotate', sample_txt)
    [
      {
        'URI': 'http://dbpedia.org/resource/Presidency_of_Barack_Obama',
        'offset': 0,
        'percentageOfSecondRank': -1.0,
        'similarityScore': 0.10031112283468246,
        'support': 134,
        'surfaceForm': 'President Obama',
        'types': 'DBpedia:OfficeHolder,DBpedia:Person,Schema:Person,Freebase:/book/book_subject,Freebase:/book,Freebase:/book/periodical_subject,Freebase:/media_common/quotation_subject,Freebase:/media_common'
      },
      …(truncated remaining elements)…
    ]

Any additional filter parameters that are supported by the Spotlight API
can be passed to the ``filters`` argument in a dictionary.

For example::

    >>> only_person_filter = {
    ...     'policy': "whitelist",
    ...     'types': "DBpedia:Person",
    ...     'coreferenceResolution': False
    ... }

    >>> spotlight.annotate(
    ...     "http://localhost/rest/annotate",
    ...     "Any collaboration between Shakira and Metallica seems highly unlikely.",
    ...     filters=only_person_filter
    ... )

    [{
        'URI': 'http://dbpedia.org/resource/Shakira',
        'offset': 26,
        'percentageOfSecondRank': 1.511934771738109e-09,
        'similarityScore': 0.9999999984880361,
        'support': 2587,
        'surfaceForm': 'Shakira',
        'types': 'Schema:MusicGroup,DBpedia:Agent,Schema:Person,DBpedia:Person,DBpedia:Artist,DBpedia:MusicalArtist'
    }]

The same parameters apply to the ``spotlight.candidates`` function,
which returns a list of all matching candidate entities rather than
only the top candidate.

Note that the Spotlight API may support other interfaces that have not been
implemented in pyspotlight. Feel free to contribute :-)!

Running DBpedia Spotlight
-----------------------------
If you just want to play around with Spotlight, there is an interactive demo
available at `demo.dbpedia-spotlight.org`_. To submit pyspotlight
requests to the demo servers, you may use the endpoints found in `sites.xml`_.

.. _demo.dbpedia-spotlight.org : http://demo.dbpedia-spotlight.org/
.. _sites.xml: http://demo.dbpedia-spotlight.org/config/sites.xml

For any significant Spotlight usage, it is strongly recommended to run
your own server. Please follow the `installation instructions`_.

.. _installation instructions: http://www.dbpedia-spotlight.org/faq#i-want-to-install-the-tool-how-do-i-do

Exceptions
----------
The following exceptions can occur:

* ``ValueError`` when:

  - the JSON response could not be decoded.

* ``SpotlightException`` when:

  - the JSON response did not contain any needed fields or was not formed as
    excepted.
  - You forgot to explicitly specify a protocol (http/https) in the API URL.

  Usually the exception's message tells you *exactly* what is wrong. If
  not, we might have forgotten some error handling. So just open up an issue on
  github if you encounter unexpected exceptions.

* ``requests.exceptions.HTTPError``

  Is thrown when the response http status code was *not* ``200``. This could happen
  if you have a load balancer like nginx in front of your spotlight cluster and
  there is not a single server available, so nginx throws a ``502 Bad Gateway``.

Tips
====

We highly recommend playing around with the *confidence* and *support* values.
Furthermore it might be preferable to filter out more annotations by looking
at their *similiarityScore* (read: contextual score).

If you want to change the default values, feel free to use ``itertools.partial``
to create a little wrapper with simplified signature::

    >>> from spotlight import annotate
    >>> from functools import partial
    >>> api = partial(annotate, 'http://localhost/rest/annotate',
    ...               confidence=0.4, support=20,
    ...               spotter='SpotXmlParser')
    >>> api('This is your test text. This function uses a non-default
    ...      confidence, support, and spotter. Furthermore all calls go
    ...      directly to localhost/rest/annotate.')

As you can see this reduces the function's complexity greatly.
Pyspotlight provides an interface based on functions rather than classes,
to avoid an unnecessary layer of indirection.

Tests
=====

If you want to run the tests, you will have to install ``nose2`` (~0.6) from PyPI.
Then you can simply run ``nose2`` from the command line in
this or the ``spotlight/`` directory.

All development and regular dependencies can be installed with a single command::

    pip install -r requirements-dev.txt


Bugs
====

In case you spot a bug, please open an issue and attach the raw response you
sent. Have a look at `ubergrape/pyspotlight#3`_ for an example on how to file a good bug report.

.. _`ubergrape/pyspotlight#3`: https://github.com/ubergrape/pyspotlight/issues/3


HISTORY.rst
Changelog
=========

v0.7.2 (2017-12-02)
-------------------

- Updated README instructions and links. [Alex Olieman]

- Ensure that ``candidates`` returns surface forms as strings. [Alex Olieman]

- Ensure that surface forms are always strings (merge `PR #1`_). [ShomyLiu & Alex Olieman]

.. _PR #1: https://github.com/aolieman/pyspotlight/pull/1

v0.7.1 (2016-07-25)
-------------------

- Moved the shared request logic in ``annotate`` and ``candidates`` to a
  helper function. [Alex Olieman]

- Updated setup/package files [Alex Olieman]

- Updated README. [Luis Nell & Alex Olieman]

v0.7.0 (2016-07-18)
-------------------

API Changes
~~~~~~~~~~~

- Changed default spotter to ``'Default'`` for 0.7 compatibility. [Alex
  Olieman]

- Moved filter parameters into a ``filters`` argument. [Alex Olieman]

  * **Removed** the ``policy`` argument from ``annotate`` and ``candidates``.
  * Added a types parameter, which enables server-side filtering of resources.
    It also makes for a nice addition to the policy parameter.

Additions
~~~~~~~~~

- Python 3 compatibility. [Alex Olieman]

- Moved to nose2 for tests. [Alex Olieman]

Fixes
~~~~~

- Updated required version of the requests package. [Alex Olieman]

- Remove mutable default arguments. [Luis Nell]

v0.6.5.2 (2013-08-27)
---------------------

- Add manifest so README is included on PyPI. [Luis Nell]

v0.6.5.1 (2013-08-12)
---------------------

- Update README for PyPI release. [Luis Nell]

- Upgrade to requests 1.2.3. [Luis Nell]

- BSD License. [Luis Nell]

- Workaround for footnotes in ``surfaceForm`` that get parsed as a list.
  [Luis Nell]

- Do not assume in ``candidates`` that ``surfaceForm`` is always a list.
  [Luis Nell]

v0.6.5 (2012-10-07)
-------------------

API Changes
~~~~~~~~~~~

- Have to explicitly provide a protocol in the URL. [Luis Nell]

Additions
~~~~~~~~~

- Added stuff for testing. [Luis Nell]

- Add requirements.txt for pip. [Luis Nell]

- Make use of requests builtin json decoding. [Luis Nell]

Fixes
~~~~~

- Some README updates. [Luis Nell]

- Add ordereddict requirement for py2.6. [Luis Nell]

- Tests: adapt to the requests raw handling. [Luis Nell]

- Use requests 0.14.1 from now on. [Luis Nell]

- Fixed typos, wrong link. [Pablo Mendes]

  * Minor: We spell it DBpedia, not DBPedia :)
  * Fix: Link pointed to OpenCalais, a commercial closed-source
    alternative to DBpedia Spotlight

v0.5.3 (2012-08-01)
-------------------

- Update README to reflect the exception changes. [Luis Nell]

- Raise requests.exceptions.HTTPError on response.status_code != 200.
  [Luis Nell]

- Prefer simplejson to json. [Luis Nell]

- Add tests for new exception handling. [Luis Nell]

- Add Exception Handling. [Luis Nell]

v0.5.2 (2012-04-06)
-------------------

- Fixes setup.py issues. v0.5.2. [Luis Nell]

v0.5.1 (2012-03-21)
-------------------

- Fix setup.py - push 0.5.1. [Luis Nell]

v0.5.0 (2012-03-20)
-------------------

- Init. [Luis Nell]


spotlight/tests.py
from collections import namedtuple, OrderedDict
from nose2.tools.such import helper
import spotlight


# Expose unittest assertions to function-style tests
assert_equals = helper.assertEquals
assert_raises = helper.assertRaises


def fake_request_post(self, *args, **kwargs):
    RawResponse = namedtuple('RawResponse', ['reason', ])
    hear_me_RawR = RawResponse(reason='Just a fake reason.')

    class FakeResponse(spotlight.requests.models.Response):
        content = kwargs['headers']['fake_response']

        def raise_for_status(self):
            self.raw = hear_me_RawR
            self.status_code = (kwargs['headers']['fake_status']
                                if 'fake_status' in kwargs['headers']
                                else spotlight.requests.codes.ok)
            return super(FakeResponse, self).raise_for_status()
    return FakeResponse()
spotlight.requests.post = fake_request_post


def test_number_convert():
    assert_equals(spotlight._convert_number('0'), 0)
    assert_equals(spotlight._convert_number('0.2'), 0.2)
    assert_equals(spotlight._convert_number(True), True)
    assert_equals(spotlight._convert_number('evi'), 'evi')
    # Testing the footnote workaround.
    assert_equals(spotlight._convert_number([1]), '[1]')


def test_protocol_missing():
    with assert_raises(spotlight.SpotlightException):
        spotlight.annotate('localhost', 'asdasdasd',
                           headers={'fake_response': b'invalid json',
                                    'fake_status': 502})


def test_http_fail():
    with assert_raises(spotlight.requests.exceptions.HTTPError):
        spotlight.annotate('http://localhost', 'asdasdasd',
                           headers={'fake_response': b'invalid json',
                                    'fake_status': 502})


def test_annotation_invalid_json():
    with assert_raises(ValueError):
        spotlight.annotate('http://localhost', 'asdasdasd',
                           headers={'fake_response': b'invalid json'})


def test_missing_resources():
    with assert_raises(spotlight.SpotlightException):
        spotlight.annotate('http://localhost', 'asdasdasd',
                           headers={'fake_response': b'{"Test": "Win"}'})


def test_candidates_invalid_json():
    with assert_raises(ValueError):
        spotlight.annotate('http://localhost', 'asdasdasd',
                           headers={'fake_response': b'invalid json'})


def test_missing_annotation():
    with assert_raises(spotlight.SpotlightException):
        spotlight.candidates('http://localhost', 'asdasdasd',
                             headers={'fake_response': b'{"Test": "Win"}'})


def test_missing_surfaceForms():
    with assert_raises(spotlight.SpotlightException):
        spotlight.candidates('http://localhost', 'asdasdasd',
                             headers={'fake_response': b'{"annotation": {"Test": "Win"}}'})


def test_single_candidate():
    # Test with a single returned candidate, as was reported by issue #3.
    # Thanks to aolieman for the awesome test data!
    data = b"""
{
   "annotation":{
      "@text":"Industrial Design at the Technische Universiteit Delft",
      "surfaceForm":{
         "@name":"Technische Universiteit Delft",
         "@offset":"25",
         "resource":[
            {
               "@label":"Technische Universiteit Delft",
               "@uri":"Technische_Universiteit_Delft",
               "@contextualScore":"0.9991813164782087",
               "@percentageOfSecondRank":"0.1422872887244497",
               "@support":"3",
               "@priorScore":"2.8799662606192636E-8",
               "@finalScore":"0.8754365122251001",
               "@types":""
            },
            {
               "@label":"Delft University of Technology",
               "@uri":"Delft_University_of_Technology",
               "@contextualScore":"8.186418452925803E-4",
               "@percentageOfSecondRank":"0.0",
               "@support":"521",
               "@priorScore":"5.001541405942121E-6",
               "@finalScore":"0.12456348777489806",
               "@types":"DBpedia:Agent, Schema:Organization, DBpedia:Organisation, Schema:EducationalOrganization, DBpedia:EducationalInstitution, Schema:CollegeOrUniversity, DBpedia:University"
            }
         ]
      }
   }
}
    """
    candidates = spotlight.candidates('http://localhost', 'asdasdasd',
                                      headers={'fake_response': data})
    expected_out = [
        {
            'resource': [
                {
                    'finalScore': 0.8754365122251001,
                    'support': 3,
                    'uri': 'Technische_Universiteit_Delft',
                    'label': 'Technische Universiteit Delft',
                    'types': '',
                    'percentageOfSecondRank': 0.1422872887244497,
                    'priorScore': 2.8799662606192636e-08,
                    'contextualScore': 0.9991813164782087
                },
                {
                    'finalScore': 0.12456348777489806,
                    'support': 521,
                    'uri': 'Delft_University_of_Technology',
                    'label': 'Delft University of Technology',
                    'types': 'DBpedia:Agent, Schema:Organization, DBpedia:Organisation, Schema:EducationalOrganization, DBpedia:EducationalInstitution, Schema:CollegeOrUniversity, DBpedia:University',
                    'percentageOfSecondRank': 0.0,
                    'priorScore': 5.001541405942121e-06,
                    'contextualScore': 0.0008186418452925803
                },
            ],
            'name': 'Technische Universiteit Delft',
            'offset': 25
        }
    ]
    assert_equals(candidates, expected_out)


def test_dict_key_cleanup():
    dirty_dict = OrderedDict()
    dirty_dict['@dirty'] = 'value'
    dirty_dict['@empty'] = None  # None values should be removed.
    dirty_dict['@recursive'] = OrderedDict()
    dirty_dict['@recursive']['tests'] = '1'
    dirty_dict['@recursive']['surfaceForm'] = '02'
    dirty_dict['@recursive']['name'] = '02'
    dirty_dict['@recursive']['stuff'] = OrderedDict()
    more = OrderedDict()
    more['something'] = 'isgoingon'
    moremore = OrderedDict()
    moremore['@moar'] = True
    moar_iterable = [more, moremore]
    dirty_dict['@recursive']['stuff'] = moar_iterable

    clean_dict = OrderedDict()
    clean_dict['dirty'] = 'value'
    clean_dict['recursive'] = OrderedDict()
    clean_dict['recursive']['tests'] = 1
    clean_dict['recursive']['surfaceForm'] = '02'
    clean_dict['recursive']['name'] = '02'
    clean_dict['recursive']['stuff'] = OrderedDict()
    more = OrderedDict()
    more['something'] = 'isgoingon'
    moremore = OrderedDict()
    moremore['moar'] = True
    moar_iterable = [more, moremore]
    clean_dict['recursive']['stuff'] = moar_iterable
    assert_equals(spotlight._dict_cleanup(dirty_dict, dict_type=OrderedDict), clean_dict)


spotlight.py
__version_info__ = (0, 5, 1)
__version__ = '.'.join(map(str, __version_info__))


try:
    import json
except ImportError:
    import simplejson as json

import requests


# Some helper functions.
def _convert_number(value):
    """
    Try to convert a string to an int or float.
    """
    if isinstance(value, bool): return value

    try:
        return int(value)
    except ValueError:
        try:
            return float(value)
        except ValueError:
            return value


def _dict_cleanup(dic, dict_type=dict):
    """
    Clean the response dictionary from ugly @ signs in keys.
    """
    clean = dict_type()
    for key, value in dic.iteritems():
        key = key.replace('@', '')
        try:
            try:
                # If this is a string or bool,
                # go straight to type conversion.
                if (isinstance(value, basestring) or
                        isinstance(value, bool)):
                    raise AttributeError
                # Test for an iterable (list, tuple, set)
                value[0]
                # Clean up each element in the iterable
                clean[key] = [_dict_cleanup(element, dict_type)
                                for element in value]
            except KeyError:
                clean[key] = _dict_cleanup(value, dict_type)
        except AttributeError:
            clean[key] = _convert_number(value)
    return clean


# Main functions.
#
# I was inspired to go back to a function based approach after seeing this
# awesome talk by Jack Diederich: Stop Writing Classes
# http://pyvideo.org/video/880/stop-writing-classes
# Most of the class-based approach had the problems he described.
# Embarrassing!
def annotate(address, text, confidence=0.0, support=0,
             spotter='LingPipeSpotter', disambiguator='Default',
             policy='whitelist', headers={}):
    """
    Annotate a text.

    :param address:
        The absolute address of the annotate REST API.
    :type address: string

    :param text:
        The text to be sent.
    :type text: string

    :param confidence:
        Filter out annotations below a given confidence.
        Based on my experience I would suggest you set this to something
        above 0.4, however your experience might vary from text to text.
    :type confidence: float

    :param support:
        Only output annotations above a given prominence (support).
        Based on my experience I would suggest you set this to something
        above 20, however your experience might vary from text to text.
    :type support: int

    :param spotter:
        One of spotters available on your DBPedia Spotlight server.
        For example one of: LingPipeSpotter, AtLeastOneNounSelector,
                            CoOccurrenceBasedSelector
    :type spotter: string

    :param disambiguator:
        The disambiguator to use on the annotation.
    :type disambiguator: string

    :param policy:
        The policy to be used.
    :type disambiguator: string

    :param headers:
        Additional headers to be set on the request.
    :type headers: dictionary

    :rtype: list of resources
    """
    payload = {'confidence': confidence, 'support': support,
               'spotter': spotter, 'disambiguator': disambiguator,
               'policy': policy, 'text': text}
    reqheaders = {'accept': 'application/json'}
    reqheaders.update(headers)
    response = requests.post(address, data=payload, headers=reqheaders)
    pydict = json.loads(response.text)
    return [_dict_cleanup(resource) for resource in pydict['Resources']]


# This is more or less a duplicate of the annotate function, with just
# the return line being the difference haha.
def candidates(address, text, confidence=0.0, support=0,
             spotter='LingPipeSpotter', disambiguator='Default',
             policy='whitelist', headers={}):
    """
    Get the candidates from a text.

    :param address:
        The absolute address of the annotate REST API.
    :type address: string

    :param text:
        The text to be sent.
    :type text: string

    :param confidence:
        Filter out annotations below a given confidence.
        Based on my experience I would suggest you set this to something
        above 0.4, however your experience might vary from text to text.
    :type confidence: float

    :param support:
        Only output annotations above a given prominence (support).
        Based on my experience I would suggest you set this to something
        above 20, however your experience might vary from text to text.
    :type support: int

    :param spotter:
        One of spotters available on your DBPedia Spotlight server.
        For example one of: LingPipeSpotter, AtLeastOneNounSelector,
                            CoOccurrenceBasedSelector
    :type spotter: string

    :param disambiguator:
        The disambiguator to use on the annotation.
    :type disambiguator: string

    :param policy:
        The policy to be used.
    :type disambiguator: string

    :param headers:
        Additional headers to be set on the request.
    :type headers: dictionary

    :rtype: list of surface forms
    """
    payload = {'confidence': confidence, 'support': support,
               'spotter': spotter, 'disambiguator': disambiguator,
               'policy': policy, 'text': text}
    reqheaders = {'accept': 'application/json'}
    reqheaders.update(headers)
    response = requests.post(address, data=payload, headers=reqheaders)
    pydict = json.loads(response.text)
    return [_dict_cleanup(form)
            for form in pydict['annotation']['surfaceForm']]


spotlight/__init__.py
"""
Python DBpedia Spotlight API Wrapper
====================================

This is just a simple interface to a Spotlight API.

Tested with DBPedia Spotlight 0.7.
"""
__version_info__ = (0, 7, 2)
__version__ = '.'.join(map(str, __version_info__))
__url__ = 'https://github.com/aolieman/pyspotlight'


import requests


class SpotlightException(Exception):
    """
    Exception raised on Spotlight failures.

    Basically this exception is raised if there was no valid JSON response
    from Spotlight.
    """
    pass


# Some helper functions.

def _post_request(address, payload, filters, headers):
    """
    Build the Spotlight request, POST it to the server, and return
    the response's JSON body.
    """
    filter_kwargs = {'policy': 'whitelist'}
    filter_kwargs.update(filters or {})
    payload.update(filter_kwargs)

    reqheaders = {'accept': 'application/json'}
    reqheaders.update(headers or {})

    # Its better for the user to have to explicitly provide a protocol in the
    # URL, since transmissions might happen over HTTPS or any other secure or
    # faster (spdy/HTTP2 :D) channel.
    if '://' not in address:
        raise SpotlightException('Oops. Looks like you forgot the protocol '
                                 '(http/https) in your url (%s).' % address)

    response = requests.post(address, data=payload, headers=reqheaders)

    # http status codes >=400,<600 shall raise an exception.
    response.raise_for_status()

    json_body = response.json()
    if json_body is None:
        raise SpotlightException("Spotlight's response did not contain valid "
                                 "JSON: %s" % response.text)

    return json_body


def _convert_number(value):
    """
    Try to convert a string to an int or float.
    """
    if isinstance(value, bool):
        return value
    # Workaround for footnotes being put into Resources.surfaceForm and then
    # having them parsed by the JSON parser into a list. (issue #4)
    if isinstance(value, list):
        value = str(value)

    try:
        return int(value)
    except ValueError:
        try:
            return float(value)
        except ValueError:
            return value


def _dict_cleanup(dic, dict_type=dict):
    """
    Clean the response dictionary from ugly @ signs in keys.

    TODO: Make this an iteration based recursion instead of function based.
          That way we can avoid stack fails.
    """
    clean = dict_type()
    for key, value in dic.items():
        if value is None:
            continue

        key = key.replace('@', '')
        try:
            try:
                # If this is a string or bool,
                # go straight to type conversion.
                if (hasattr(value, 'strip') or
                        isinstance(value, bool)):
                    raise AttributeError
                # Test for an iterable (list, tuple, set)
                value[0]
                # Clean up each element in the iterable
                clean[key] = [
                    _dict_cleanup(element, dict_type)
                    for element in value
                ]
            except KeyError:
                clean[key] = _dict_cleanup(value, dict_type)
        except AttributeError:
            if key in {'surfaceForm', 'name'}:
                clean[key] = value
            else:
                clean[key] = _convert_number(value)
    return clean


# Main functions.

def annotate(address, text, confidence=0.0, support=0,
             spotter='Default', disambiguator='Default',
             filters=None, headers=None):
    """
    Get semantic annotations (i.e. entity links) from a text.

    Can raise :exc:`requests.exceptions.HTTPError` or
    :exc:`SpotlightException`, depending on where the failure is (HTTP status
    code not 200 or the response not containing valid json).

    :param address:
        The absolute address of the annotate REST API.
    :type address: string

    :param text:
        The text to be sent.
    :type text: string

    :param confidence:
        Filter out annotations below a given confidence.
        Based on my experience I would suggest you set this to something
        above 0.4, however your experience might vary from text to text.
    :type confidence: float

    :param support:
        Only output annotations above a given prominence (i.e. support,
        indegree on Wikipedia).
    :type support: integer

    :param spotter:
        One of spotters available on your DBPedia Spotlight server.
        For example one of: LingPipeSpotter, AtLeastOneNounSelector,
                            CoOccurrenceBasedSelector
    :type spotter: string

    :param disambiguator:
        The disambiguator to use on the annotation.
    :type disambiguator: string

    :param filters:
        Additional parameters that collectively define a filter function.

        For example:
        'policy'                (string)
                                The policy to be used:
                                'whitelist' or 'blacklist';
        'types'                 (string)
                                Comma-separated list of types,
                                i.e. 'DBpedia:Agent,Schema:Organization';
        'sparql'                (string)
                                Select only entities that (don't)
                                match with the SPARQL query result;
        'coreferenceResolution' (boolean)
                                Annotate coreferences: true / false.
                                Set to false to use types (statistical only).

    :type filters: dictionary

    :param headers:
        Additional headers to be set on the request.
    :type headers: dictionary

    :rtype: list of resources
    """
    payload = {
        'confidence': confidence,
        'support': support,
        'text': text,
        'spotter': spotter,
        'disambiguator': disambiguator
    }

    pydict = _post_request(address, payload, filters, headers)

    if 'Resources' not in pydict:
        raise SpotlightException(
            'No Resources found in spotlight response: %s' % pydict
        )

    return [_dict_cleanup(resource) for resource in pydict['Resources']]


def candidates(address, text, confidence=0.0, support=0,
               spotter='Default', disambiguator='Default',
               filters=None, headers=None):
    """
    Get the candidate entities from a text.

    Uses the same arguments as :meth:`annotate`.

    :rtype: list of surface forms
    """
    payload = {
        'confidence': confidence,
        'support': support,
        'text': text,
        'spotter': spotter,
        'disambiguator': disambiguator
    }

    pydict = _post_request(address, payload, filters, headers)

    if 'annotation' not in pydict:
        raise SpotlightException(
            'No annotations found in spotlight response: %s' % pydict
        )
    if 'surfaceForm' not in pydict['annotation']:
        raise SpotlightException(
            'No surface forms found in spotlight response: %s' % pydict
        )

    # Previously we assumed that the surfaceForm is *always* a list, however
    # depending on how many are returned, this does not have to be the case.
    # So we are doing some good ol' duck typing here.
    try:
        pydict['annotation']['surfaceForm'][0]
    except KeyError:
        # However note that we will *always* return a list.
        return [_dict_cleanup(pydict['annotation']['surfaceForm']), ]
    return [
        _dict_cleanup(form)
        for form in pydict['annotation']['surfaceForm']
    ]
